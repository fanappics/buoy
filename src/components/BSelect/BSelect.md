```js

const options = [{id: 1, displayText: 'This'}, {id: 2, displayText: 'Is'}, {id: 3, displayText: 'An Example'}]
const selected1 = null
const selected2 = null
const selected3 = 3

<div>
  <div style="width: 50%">
    <b-select
      :selectOptions="options"
      id="example-id"
      placeholder="Placeholder Example"
      selectLabel="Select Example"
      v-model="selected1"
    />
  </div>

  <div>
    <p>Selected: {{ selected1 }}</p>
  </div>
  
  <div style="width: 50%">
    <b-select
      :selectOptions="options"
      id="example-id-2"
      placeholder="Placeholder Example"
      required
      selectLabel="Required Select Example"
      v-model="selected2"
    />
  </div>

  <div>
    <p>Selected: {{ selected2 }}</p>
  </div>
  
  <div style="width: 50%">
    <b-select
      :selectOptions="options"
      id="example-id-3"
      required
      selectLabel="Pre Selected Example"
      v-model="selected3"
    />
  </div>

  <div>
    <p>Selected: {{ selected3 }}</p>
  </div>

</div>
```