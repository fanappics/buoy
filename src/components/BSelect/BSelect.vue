<template>
  <div>
    <div>
      <label>{{ selectLabel }} <span v-if=isRequired aria-label='Required'>*</span></label>
    </div>
    <div 
      :aria-expanded="opened"
      :aria-owns="id"
      :aria-label="selectLabel"
      :aria-required="isRequired"
      role="combobox"
      tabindex="0"
      v-model="selectLabel"
      :id="'dropdown' + id"
      @keyup="keyUpHandler(selectedOption)"
    >
      <ul
        :class="ulClass"
        role="listbox"
        :id="id"
      >
        <li 
          :class="dropdownClass" 
          @click="toggleList"
          >{{ selectedOption ? selectedOption.text : placeholder }}
          <span class="dropdown-arrow"><i class="fa fa-caret-down" aria-hidden="true"></i></span>
        </li>
        <li
          v-if=opened
          v-for="option in options"
          class="li-opened options"
          :key="'selectOption' + option.id"
          :id="'option' + option.id"
          :value=option.id
          :aria-selected="selectedOption && selectedOption.id === option.id"
          @click="selectOption(option)"
          role="option"
          :ref="'option' + option.id"
          @keyup="keyUpHandler(option)"
          tabindex='-1'
        > {{ option.text }}
        </li>
      </ul>
    </div>
  </div>
</template>

<script>

const options = []

export default {
  name: 'b-select',
  props: {

    id: {
      type: String,
      required: true
    },
    selectOptions: {
      type: Array
    },
    isRequired: {
      type: Boolean,
      required: true
    },
    selectLabel: {
      type: String,
      required: true
    },
    initialValue: {
      type: Object,
      required: false
    },
    placeholder: {
      type: String,
      required: false,
      default: 'Choose Option'
    }
  },

  methods: {

    /**
     * Shows and hides line options for the select list.
     */

    toggleList () {
      this.opened = !this.opened
      if (this.opened) {
        this.$nextTick( function () {
          this.setFocus()
        })
      } else {
        document.querySelector(`#dropdown${this.id}`).focus()
      }
    },

    /**
     * @param {Object} option
     * Handles the selection of a line and closed the list.
     */

    selectOption (option) {
      this.selectedOption = option
      this.opened = !this.opened
      document.querySelector(`#dropdown${this.id}`).focus()
    },

    /**
     * Handles the focus setting for accessibility purposes
     */

    setFocus () {
      if (this.selectedOption) {
        const line = document.querySelector(`#${this.id} #option${this.selectedOption.id}`)
        line.focus()
      } else {
        const line = document.querySelector(`#${this.id} #option${this.options[0].id}`)
        line.focus()  
      }
    },

    /**
     * @param {object} option
     * Handles keypresses for keyboard navigation in the component.
     * Up arrow (38), down arrow (40), space bar (32), and return (13) are supported.
     */

    keyUpHandler (option) {
      const target = event.target;
      event.stopPropagation();
      if(event.keyCode === 40) {
        event.preventDefault();
        if(target.nextElementSibling) {
          const next = target.nextElementSibling;
          next.setAttribute("tabindex", "0");
          target.setAttribute("tabindex", "-1");
          next.focus();
        }
      } else if(event.keyCode === 38) {
        event.preventDefault();
        if(target.previousElementSibling) {
          const next = target.previousElementSibling;
          const first = next.previousElementSibling;
          if (first) {
            next.setAttribute("tabindex", "0");
            target.setAttribute("tabindex", "-1");
            next.focus();
          }
        }
      } else if ( event.keyCode === 32 && !this.opened) {
        event.preventDefault();
        this.toggleList()
      } else if (event.keyCode === 32 || event.keyCode === 13) {
        if (this.opened) {
          event.preventDefault();
          this.selectOption(option)
        }
      }
    }
  },

  data () {
    return {
      options: this.selectOptions,
      opened: false,
      selectedOption: this.initialValue
    }
  },

  computed: {
    ulClass: function () {
      return {
        'ul-opened': this.opened
      }
    },
    dropdownClass: function () {
      return {
        'li-opened': this.opened,
        'options': true,
        'selected': this.selectedOption,
        'unselected': !this.selectedOption
      }
    }
  }
}

</script>

<style scoped>
  .options {
    width: 100%;
    padding: .5rem;
    border-style: solid;
    border-radius: .25rem;
    border-width: thin;
    border-color: #cccccc;
  }

  ul {
    padding: 0 1rem 0 0;
    width: 100%;
    list-style: none;
  }

  .ul-opened {
    border-style: solid;
    border-radius: .25rem;
    border-width: thin;
    box-shadow:  .125rem .25rem .3rem #cccccc;
    border-color: #cccccc;
  }

  .li-opened {
    border-radius: 0;
    border-style: none none solid none;
    border-width: thin;

  }  

  .li-opened[aria-selected] {
    font-weight: 600;
  }

  .li-opened:hover:not(:first-child) {
    background-color: #f2f2f2;
  }

  .li-opened:last-child {
    border-style: none;
  }

  .dropdown-arrow {
    float: right;
    color: #000000
  }

  .selected {
    color: #000000;
  }

  .unselected {
    color: #cccccc;
  }

</style>
