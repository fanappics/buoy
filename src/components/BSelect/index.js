import BSelect from './BSelect.vue';

/* istanbul ignore next */
BSelect.install = function install (Vue) {
  Vue.component(BSelect.name, BSelect);
};

export default BSelect;
