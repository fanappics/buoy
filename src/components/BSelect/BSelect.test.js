import { shallow, mount } from 'vue-test-utils'
import BSelect from './BSelect'

describe('BSelect', () => {
  const selectOptions = [{id: 1, text: 'test'}]
  const initialValue = {id: 1, text: 'test'}
  const isRequired = false
  const selectLabel = 'test'
  const id = 'testId'
  const placeholder = 'test placeholder'

  it('creates options when passed', () => {
    const wrapper = shallow(BSelect, {
      propsData: {
        id: id,
        selectOptions: selectOptions,
        isRequired: isRequired,
        selectLabel: selectLabel
      }
    })
    expect(wrapper.props().selectOptions.length).toBe(selectOptions.length)
  })

  it('intializes selected value when passed', () => {
    const wrapper = mount(BSelect, {
      propsData: {
        id: id,
        selectOptions: selectOptions,
        isRequired: isRequired,
        selectLabel: selectLabel,
        initialValue: initialValue
      }
    })
    const initialLine = wrapper.find('li')
    expect(initialLine.text()).toBe('test')
  })

  it('display placeholder if no initial value', () => {
    const wrapper = mount(BSelect, {
      propsData: {
        id: id,
        selectOptions: selectOptions,
        isRequired: isRequired,
        selectLabel: selectLabel,
        placeholder: placeholder
      }
    })
    const initialLine = wrapper.find('li')
    expect(initialLine.text()).toBe('test placeholder')
  })

  it('opens list when line is clicked', () => {
    const wrapper = mount(BSelect, {
      propsData: {
        id: id,
        selectOptions: selectOptions,
        isRequired: isRequired,
        selectLabel: selectLabel,
        placeholder: placeholder
      }
    })
    const initialLine = wrapper.find('li')
    expect(initialLine.hasClass('li-opened')).toBe(false)
    initialLine.trigger('click')
    expect(initialLine.hasClass('li-opened')).toBe(true)
  })

  it('rendors a default select list', () => {
    const wrapper = mount(BSelect, {
      propsData: {
        id: id,
        selectOptions: selectOptions,
        isRequired: isRequired,
        selectLabel: selectLabel
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('rendors a select list with an initial value', () => {
    const wrapper = mount(BSelect, {
      propsData: {
        id: id,
        selectOptions: selectOptions,
        isRequired: isRequired,
        selectLabel: selectLabel,
        initialValue: initialValue
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('rendors a select list with a placeholder', () => {
    const wrapper = mount(BSelect, {
      propsData: {
        id: id,
        selectOptions: selectOptions,
        isRequired: isRequired,
        selectLabel: selectLabel,
        placeholder: placeholder
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
