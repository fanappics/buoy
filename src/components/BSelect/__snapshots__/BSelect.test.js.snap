// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BSelect rendors a default select list 1`] = `
<div>
    <div>
        <label>test
            <!---->
        </label>
    </div>
    <div aria-label="test" aria-owns="testId" id="dropdowntestId" role="combobox" tabindex="0">
        <ul id="testId" role="listbox" class="">
            <li class="options unselected">Choose Option
                <span class="dropdown-arrow"><i aria-hidden="true" class="fa fa-caret-down"></i></span></li>
            <!---->
        </ul>
    </div>
</div>
`;

exports[`BSelect rendors a select list with a placeholder 1`] = `
<div>
    <div>
        <label>test
            <!---->
        </label>
    </div>
    <div aria-label="test" aria-owns="testId" id="dropdowntestId" role="combobox" tabindex="0">
        <ul id="testId" role="listbox" class="">
            <li class="options unselected">test placeholder
                <span class="dropdown-arrow"><i aria-hidden="true" class="fa fa-caret-down"></i></span></li>
            <!---->
        </ul>
    </div>
</div>
`;

exports[`BSelect rendors a select list with an initial value 1`] = `
<div>
    <div>
        <label>test
            <!---->
        </label>
    </div>
    <div aria-label="test" aria-owns="testId" id="dropdowntestId" role="combobox" tabindex="0">
        <ul id="testId" role="listbox" class="">
            <li class="options selected">test
                <span class="dropdown-arrow"><i aria-hidden="true" class="fa fa-caret-down"></i></span></li>
            <!---->
        </ul>
    </div>
</div>
`;
